UefiTestingPkg/AuditTests/PagingAudit/UEFI/Dxe/App/DxePagingAuditTestApp.c

```diff
@@ -14,95 +14,209 @@ SPDX-License-Identifier: BSD-2-Clause-Patent
 #include <Protocol/SimpleFileSystem.h>
 #include <Protocol/MemoryProtectionSpecialRegionProtocol.h>
 #include <Protocol/MemoryProtectionDebug.h>
+#include <Protocol/MemoryAttribute.h>
 
 #include <Library/FileHandleLib.h>
 #include <Library/DxeServicesTableLib.h>
 #include <Library/FlatPageTableLib.h>
 #include <Library/UnitTestLib.h>
+#include <Library/HobLib.h>
+#include <Library/SafeIntLib.h>
 
 #define UNIT_TEST_APP_NAME     "Paging Audit Test"
-#define UNIT_TEST_APP_VERSION  "1"
-#define MAX_CHARS_TO_READ      3
+#define UNIT_TEST_APP_VERSION  "2"
+#define MAX_CHARS_TO_READ      4
 
 // TRUE if A interval subsumes B interval
 #define CHECK_SUBSUMPTION(AStart, AEnd, BStart, BEnd) \
   ((AStart <= BStart) && (AEnd >= BEnd))
 
-CHAR8                             *mMemoryInfoDatabaseBuffer   = NULL;
-UINTN                             mMemoryInfoDatabaseSize      = 0;
-UINTN                             mMemoryInfoDatabaseAllocSize = 0;
-MEMORY_PROTECTION_SPECIAL_REGION  *mSpecialRegions             = NULL;
-IMAGE_RANGE_DESCRIPTOR            *mNonProtectedImageList      = NULL;
-UINTN                             mSpecialRegionCount          = 0;
-EFI_GCD_MEMORY_SPACE_DESCRIPTOR   *mMemorySpaceMap             = NULL;
-UINTN                             mMemorySpaceMapCount         = 0;
-PAGE_MAP                          mMap                         = { 0 };
+// TRUE if A and B have overlapping intervals
+#define CHECK_OVERLAP(AStart, AEnd, BStart, BEnd)   \
+  ((AEnd > AStart) && (BEnd > BStart) &&            \
+  ((AStart <= BStart && AEnd > BStart) ||           \
+  (BStart <= AStart && BEnd > AStart)))
+
+// Aligns the input address down to the nearest page boundary
+#define ALIGN_ADDRESS(Address)  ((Address / EFI_PAGE_SIZE) * EFI_PAGE_SIZE)
+
+// Globals required to create the memory info database
+CHAR8  *mMemoryInfoDatabaseBuffer   = NULL;
+UINTN  mMemoryInfoDatabaseSize      = 0;
+UINTN  mMemoryInfoDatabaseAllocSize = 0;
+
+// Globals for memory protection special regions
+MEMORY_PROTECTION_SPECIAL_REGION  *mSpecialRegions    = NULL;
+UINTN                             mSpecialRegionCount = 0;
+
+// Global for the non-protected image list
+IMAGE_RANGE_DESCRIPTOR  *mNonProtectedImageList = NULL;
+
+// Globals for the memory space map
+EFI_GCD_MEMORY_SPACE_DESCRIPTOR  *mMemorySpaceMap     = NULL;
+UINTN                            mMemorySpaceMapCount = 0;
+
+// Globals for the EFI memory map
+UINTN                  mEfiMemoryMapSize           = 0;
+EFI_MEMORY_DESCRIPTOR  *mEfiMemoryMap              = NULL;
+UINTN                  mEfiMemoryMapDescriptorSize = 0;
+
+// Global for the flat page table
+PAGE_MAP  mMap = { 0 };
+
+// -------------------------------------------------
+//    GLOBALS SUPPORT FUNCTIONS
+// -------------------------------------------------
 
 /**
-  Frees the entries in the mMap global.
+  Return if the PE image section is aligned.

```

Comment:
It would be nice to mention in the function description that this will assert if data types memory types are given.

Reply:
Description updated to note that the function should only be called with the image code memory type otherwise it will ASSERT.